aspnetcore/blazor/webassembly-native-dependencies.md

```diff
@@ -0,0 +1,159 @@
+---
+title: ASP.NET Core Blazor WebAssembly native dependencies
+author: guardrex
+description: Learn how to build Blazor WebAssembly apps with native dependencies built to run on WebAssembly in the browser.
+monikerRange: '>= aspnetcore-6.0'
+ms.author: riande
+ms.custom: mvc
+ms.date: 10/28/2021
+no-loc: [Home, Privacy, Kestrel, appsettings.json, "ASP.NET Core Identity", cookie, Cookie, Blazor, "Blazor Server", "Blazor WebAssembly", "Identity", "Let's Encrypt", Razor, SignalR]
+uid: blazor/webassembly-native-dependencies
+---
+# ASP.NET Core Blazor WebAssembly native dependencies
+
+*This article applies to Blazor WebAssembly.*
+
+Blazor WebAssembly apps can use native dependencies built to run on WebAssembly. You can statically link native dependencies into the .NET WebAssembly runtime using the .NET WebAssembly build tools, the same tools used to [ahead-of-time (AOT) compile](xref:blazor/host-and-deploy/webassembly#ahead-of-time-aot-compilation) a Blazor app to WebAssembly and to [relink the runtime to remove unused features](xref:blazor/host-and-deploy/webassembly#runtime-relinking).
+
+## .NET WebAssembly build tools
+
+The .NET WebAssembly build tools are based on [Emscripten](https://emscripten.org/), a compiler toolchain for the web platform. To install the .NET WebAssembly build tools, use either of the following approaches:
+
+* Select the optional component in the Visual Studio installer.
+* Run `dotnet workload install wasm-tools` from an administrative command prompt.
+
+Add native dependencies to a Blazor WebAssembly app by adding `NativeFileReference` items in the app's project file. When the project is built, each `NativeFileReference` is passed to Emscripten by the .NET WebAssembly build tools so that they are compiled and linked into the runtime. Next, [`p/invoke`](/dotnet/standard/native-interop/pinvoke) into the native code from the app's .NET code.
+
+Generally, any portable native code can be used as a native dependency with Blazor WebAssembly. You can add native dependencies to C/C++ code or code previously compiled using Emscripten:
+
+* Object files (`.o`)
+* Archive files (`.a`)
+* Bitcode (`.bc`)
+* Standalone WebAssembly modules (`.wasm`)
+
+Prebuilt dependencies typically must be built using the same version of Emscripten used to build the .NET WebAssembly runtime.
+
+## Use native code
+
+Add a simple native C function to a Blazor WebAssembly app:
+
+1. Create a new Blazor WebAssembly project.
+1. Add the `WasmBuildNative` property to a property group in the app's project file (`.csproj`):
+
+   ```xml
+   <WasmBuildNative>true</WasmBuildNative>
```

Comment:
You don't need this. As long as there's at least one `NativeFileReference`, it will do the right thing anyway.